File: leetcode/rotate_array/soln.py
import math
from typing import List

class Solution:
    def rotate_even(self, nums, k) -> None:
        start = 0
        cur_ind = 0
        hold = nums[0]
        total_changed = 0
        len_nums = len(nums)
        first_flag = True
        while total_changed < len_nums:
            while first_flag or cur_ind != start:
                new_ind = (cur_ind + k) % len_nums
                nums[new_ind], hold = hold, nums[new_ind]
                cur_ind = new_ind
                total_changed += 1
                first_flag = False
            start += 1
            cur_ind = start
            hold = nums[cur_ind]
            first_flag = True
        return nums


    def __call__(self, n, k):
        return self.rotate_even(n, k)

File: leetcode/rotate_array/soln.cpp
#include <vector>
using namespace std;

class Solution {
public:
    void rotate(vector<int>& nums, int k) {
        int start = 0;
        int len_nums = nums.size();
        int total_counted = 0;
        int cur_idx = start;
        int next_idx = (cur_idx + k) % len_nums;
        int hold = nums[cur_idx];
        while (total_counted < len_nums) {
            while (true) {
                next_idx = (cur_idx + k) % len_nums;

                int temp_hold = nums[next_idx];
                nums[next_idx] = hold;
                hold = temp_hold;
                cur_idx = next_idx;
                total_counted++;
                /*
                for (auto i : nums) {
                    std::cout << i << " ";   
                }
                std:: cout << "\n";*/
                if (start == next_idx) {
                    break;   
                }
            }
            start++;
            cur_idx = start;
            hold = nums[cur_idx];
        }
    }
};
